wpilib/wpilib/shuffleboard/container.py

```diff
@@ -0,0 +1,108 @@
+# validated: 2018-11-18 EN 175c6c1f0130 edu/wpi/first/wpilibj/shuffleboard/ContainerHelper.java
+from networktables.entry import NetworkTableEntry
+from .._impl.utils import match_arglist
+from ..sendable import Sendable
+
+__all__ = ['ShuffleboardContainer']
+
+class ShuffleboardContainer:
+    """Common interface for objects that can contain shuffleboard components."""
+    def __init__(self):
+        self.usedTitles = set()
+        self.components = []
+        self.layouts = {}
+
+    def getComponents(self):
+        """Gets the components that are direct children of this container."""
+        return self.components
+
+    def getLayout(self, type: str, title: str):
+        """
+        Gets the layout with the given type and title, creating it if it does 
+        not already exist at the time this method is called.
+
+        :param type:  the type of the layout, eg "List" or "Grid"
+        :param title: the title of the layout
+        :returns: the layout
+        """
+        from .layout import ShuffleboardLayout
+        if title not in self.layouts:
+            layout = ShuffleboardLayout(self, type, title)
+            self.components.append(layout)
+            self.layouts[title] = layout
+
+        return self.layouts[title]
+
+    def add(self, *args, **kwargs) -> 'ShuffleboardWidget':
+        """
+        Adds a widget to this container to display the given sendable.
+
+        raises AssertionError if a widget already exists in this container 
+        with the given title
+
+        :param title:    the title of the widget. if not provided, `sendable.getName()` is used.
+        :param sendable: the sendable to display
+        :param defaultValue: the default value of the widget. not to be combined with `sendable`
+        :returns: a widget to display the sendable data          
```

Comment:
Should add a python-specific note that empty lists are not usable here.

Reply:
in the default value?

Reply:
why not?

Reply:
Because it wouldn't be possible to know the type of the list.

Reply:
that's kind of an annoying restriction. would it make sense to have networktables accept something like typing.List[str] to signal you want an empty list of strings?

Reply:
There's no way to uniquely represent an empty value of that type ­— the type of the elements could be anything.

Reply:
In theory we could make simple wrappers around list:

```
class StringList(list):
    pass
```

... but that seems more trouble than it's worth. In python, users can just use putXXXArray if they absolutely must put an empty array of things in.